src/host.ps1

```diff
@@ -1,43 +1,73 @@
-            Power = $InputObject.power_state
-            StartTime = $InputObject.startTime
-            Tags = $InputObject.tags
-            Address = $InputObject.address
-            Bios = $InputObject.bios_strings.bios_vendor
-            CPUModel = $InputObject.cpu_info.modelname
-            Build = $InputObject.software_version.build_number
-            Version = $InputObject.software_version.product_version
-            ProductBrand = $InputObject.software_version.product_brand
+    if ($HostData -is [string]) {
+        try {
+            Write-Verbose "HostData is a string, attempting to convert from JSON"
+            $HostData = $HostData | ConvertFrom-Json -AsHashtable
+        } catch {
+            Write-Error "Failed to convert host data from JSON string: $_"
+            return $null
+        }
+    }
+    
+    Write-Verbose "Processing host data: name_label=$($HostData.name_label), id=$($HostData.id), uuid=$($HostData.uuid), address=$($HostData.address)"
+
+    $hostObj = [PSCustomObject]@{
+        PSTypeName  = "XoHost"
+        HostUuid    = $HostData.id ?? $HostData.uuid ?? ""
+        Name        = $HostData.name_label ?? "Unknown"
+        Address     = $HostData.address ?? "Unknown"
+        PowerState  = $HostData.power_state ?? "Unknown"
+        MemoryFree  = 0
+        MemoryTotal = 0
+        Cpus        = 0
+        Description = $HostData.name_description
+        StartTime   = $HostData.startTime
+        Tags        = $HostData.tags
+        Version     = $HostData.version
+        ProductBrand = $HostData.productBrand
+    }
+
+    if ($HostData.ContainsKey('cpus') -and $HostData['cpus'] -and $HostData['cpus'].ContainsKey('cores')) {
+        $hostObj.Cpus = $HostData['cpus']['cores']
+    } elseif ($HostData.ContainsKey('CPUs') -and $HostData['CPUs'] -and $HostData['CPUs'].ContainsKey('cpu_count')) {
+        $hostObj.Cpus = [int]$HostData['CPUs']['cpu_count']
+    }
+
+    if ($HostData.ContainsKey('memory') -and $HostData['memory']) {
+        $memory = $HostData['memory']
+        if ($memory.ContainsKey('size')) {
+            $hostObj.MemoryTotal = [math]::Round($memory['size'] / 1GB, 2)
+        }
+        if ($memory.ContainsKey('usage')) {
+            $memoryFree = $memory['size'] - $memory['usage']
+            $hostObj.MemoryFree = [math]::Round($memoryFree / 1GB, 2)
         }
-        
-        return $resultObject
     }
+
+    if ($hostObj.PowerState -and $hostObj.PowerState -ne "Unknown") {
+        $hostObj.PowerState = $hostObj.PowerState.Replace("_", " ")
+        $hostObj.PowerState = (Get-Culture).TextInfo.ToTitleCase($hostObj.PowerState.ToLower())
+    }
+
+    Write-Verbose "Created host object: Name=$($hostObj.Name), UUID=$($hostObj.HostUuid), Address=$($hostObj.Address), PowerState=$($hostObj.PowerState), MemoryTotal=$($hostObj.MemoryTotal), MemoryFree=$($hostObj.MemoryFree), Cpus=$($hostObj.Cpus)"
+    return $hostObj
 }
 
 function Get-XoHost {
```

Comment:
Will doing everything in `end{}` break pipelines? Please stick with one cmdlet structure in all of your request (`Get-Xo...`) cmdlets.